File: 1277_count_squares/count_squares.py
from typing import List


def count_squares(matrix: List[List[int]]) -> int:
    """
    Time: O(m * n)
    Space: O(1)
    """
    res = 0
    for r in range(len(matrix)):
        for c in range(len(matrix[0])):
            if r and c and matrix[r][c]:
                matrix[r][c] = min(matrix[r - 1][c], matrix[r]
                                   [c - 1], matrix[r - 1][c - 1]) + 1
            res += matrix[r][c]
    return res

File: 1277_count_squares/count_squares.go
package main

import "math"

/**
 * Time: O(m * n)
 * Space: O(1)
 */
func countSquares(matrix [][]int) int {
	res := 0
	for r := 0; r < len(matrix); r++ {
		for c := 0; c < len(matrix[0]); c++ {
			if r > 0 && c > 0 && matrix[r][c] == 1 {
				matrix[r][c] = min(matrix[r-1][c], matrix[r][c-1], matrix[r-1][c-1]) + 1
			}
			res += matrix[r][c]
		}
	}
	return res
}

func min(nums ...int) int {
	res := math.MaxInt16
	for _, num := range nums {
		i := num
		if i < res {
			res = i
		}
	}
	return res
}

func main() {

}
